firedrake/interpolation.py

```diff
@@ -279,27 +288,22 @@ def _interpolator(V, tensor, expr, subset, arguments, access):
     # FIXME: for the runtime unknown point set (for cross-mesh
     # interpolation) we have to pass the finat element we construct
     # here. Ideally we would only pass the UFL element through.
-    kernel = compile_expression_dual_evaluation(expr, to_element,
-                                                V.ufl_element(),
-                                                domain=source_mesh,
-                                                parameters=parameters)
+    kernel = compile_expression(cell_set.comm, expr, to_element, V.ufl_element(),
+                                domain=source_mesh, parameters=parameters)
     ast = kernel.ast
     oriented = kernel.oriented
     needs_cell_sizes = kernel.needs_cell_sizes
-    coefficients = kernel.coefficients
+    coefficient_numbers = kernel.coefficient_numbers
     first_coeff_fake_coords = kernel.first_coefficient_fake_coords
     name = kernel.name
     kernel = op2.Kernel(ast, name, requires_zeroed_output_arguments=True,
                         flop_count=kernel.flop_count)
-    cell_set = target_mesh.cell_set
-    if subset is not None:
-        assert subset.superset == cell_set
-        cell_set = subset
     parloop_args = [kernel, cell_set]
 
+    coefficients = tsfc_interface.extract_numbered_coefficients(expr, coefficient_numbers)
     if first_coeff_fake_coords:
         # Replace with real source mesh coordinates
-        coefficients[0] = source_mesh.coordinates
+        coefficients = [source_mesh.coordinates] + coefficients
```

Comment:
I think `first_coeff_fake_coords` needs to be renamed here (and perhaps elsewhere). It no longer tells you to replace the coordinates coefficient, instead you are adding a set of source mesh coordinates which aren't otherwise in the coefficient list.

Reply:
At the very least this comment
```
# Replace with real source mesh coordinates
```
needs to change since the addition of
```
coefficients = tsfc_interface.extract_numbered_coefficients(expr, coefficient_numbers)
```
means that's not what you are doing any more.

Reply:
`needs_external_coords` seems fine. Does it make logical sense to name it that in TSFC too, or has the API not been changed there sufficiently to warrant it?

Reply:
I think it is warranted since interpolation kernels no longer advertise a `coefficients` property and hence 'first coefficient' (from `first_coeff_fake_coords`) is misleading.

Reply:
This is now done.